Implement internationalization support

msujew opened the pull request:
#### What it does

Closes #1327

Adds internationalization support for Theia. The following features are supported:

- [x] Support for [vscode language packs](https://code.visualstudio.com/docs/getstarted/locales) to localize Theia itself.
- [x] Support for localized vscode extensions (both `package.json` and runtime translations via `nls.localize()`)
- [x] Support for custom Theia localizations through a `LocalizationContribution` interface.
- [x] Support for localization of the monaco editor (search widget, context menu and so on).
- [x] Showing the original text under the localized value (for commands).

Things that are out of scope for this PR (and will probably be addressed in another PR):

- Support for vscode language packs to localize other vscode extensions (like the builtins)

#### How it works

1. The `core/browser` package contains a `nls` namespace with a `localize` function similiar to vscode. It accepts a translation key, a default value (usually the original english value) and additional args. The translations are loaded before any other imports so that they are available during load-time of any other code.
2. The backend application loads vscode language packs and Theia's `LocalizationContribution`s and places them in a `LocalizationProvider`. Additionally, the backend exposes a `/i18n/<locale>` HTTP endpoint to retrieve localizations for any locale.
3. The vscode plugin host starts the vscode extension process with the locale that is currently used by the user. This allows vscode extensions to translate themselves.
4. The `monaco-loader.ts` uses the frontend locale as well to configure monaco to use the currently active language.

#### How to test

0. Test that the `Configure display language` command can be invoked but only displays `en` (even though german translations exist in the backend, but no language pack is installed)
1. Download & install the german language pack (v1.50, search for `german` in the Theia extension menu)
2. Download & install [the i18n sample extension in german](https://github.com/eclipse-theia/theia/files/6658579/i18n-sample-0.1.0.zip)
3. Execute the `Hello` and `Bye` command contributed by the i18n sample extension. Those will simply show `Hello` and `Bye` respectively in a notification.
4. Execute the `Configure Display Language` command and choose the german locale `de`.
5. The application will be reloaded (you will maybe have to reload it yourself) and partially translated into german. You will see this in the main menu as well as the workspace and some common commands. *(Test for vscode localization packs)*
6. Execute the `Hallo` (Hello) and `Auf Wiedersehen` (Bye) command, which will now display german text as a notification. *(Test for vscode extensions to correctly localize themselves and for Theia to read the translated `package.json` correctly)*
7. In the context menu of the file explorer, you will see that the `Download` menu item has been replaced with `Herunterladen`. This was done with a german `LocalizationContribution`.
8. (optionally) Replace the `editor.main.nls.de.js` in the `examples/browser/lib/vs/editor` source directory with [this file](https://github.com/eclipse-theia/theia/files/6577192/editor.main.nls.de.zip). Run `gzip -k editor.main.nls.de.js` in that directory. Reloading the Theia application and using `de` locale will now show german translation in the monaco search widget (`Ctrl+F`) as well as in the context menu. *This is a workaround, see below for more information.*

#### Open Issues

* Do we need internationalization support for numbers and dates? - _Will be addressed in a separate PR, if at all_
* How about right to left languages?
* The `@theia/monaco-editor-core` package contains only english translations. Is this done on purpose? Replacing the `nls.*.json` files with the correct translations will translate monaco correctly.
* The translations for the monaco editor rely on the order of certain language pack versions. This isn't an issue for the default languages (like `de`, `it`, `fr`...), but makes it virtually impossible for vscode language packs to actually contribute to monaco's translation process.
* Language Packs can't be uninstalled without restarting the server (all localizations are held in memory after startup), as we don't know which translations are contributed by vscode language packs and which by Theia extensions **yet**.
* Currently the `LocalizationProvider` holds a single active language for the whole application. Is there a way to hold a value for every user?
* How do we handle translations contributed by Theia extensions without having a language pack installed? Selecting the german locale e.g. will just translate Theia's own commands, not the rest.
* Theia's quick-input-bar doesn't filter by the details of the item, unlike vscode. This makes finding commands quite difficult after switching the locale.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work commented on packages/keymaps/src/browser/keybindings-widget.tsx:
```diff
@@ -492,8 +496,9 @@ export class KeybindingWidget extends ReactWidget {
+            const label = LocalizationInfo.localize(command.label, this.localizationService);
             // Prefix the command label with the category if it exists, else return the simple label.
-            return command.category ? `${command.category}: ${command.label}` : command.label;
+            return command.category ? `${command.category}: ${label}` : label;
```
It looks like command categories are also available for localization in VSCode:
![image](https://user-images.githubusercontent.com/62660806/120397918-d4a50b80-c2fe-11eb-9e22-b8cb376212ad.png)

Would it be possible to add those to the PR conveniently?

msujew replied:
Totally forgot about those. Done (for the 'File' command entries at least)

Also, you reminded me that vscode shows the original text under the localized text. Added this feature to the planned features for this PR.

colin-grant-work commented:
A general observation: as this is currently implemented, a lot of services need to know about it or opt into it. For example:
``` typescript
    export function compareCommands(a: Command, b: Command): number {
        if (a.label && b.label) {
            const aLabel = LocalizationInfo.localize(a.label);
            const bLabel = LocalizationInfo.localize(b.label);
            const aCommand = (a.category ? `${LocalizationInfo.localize(a.category)}: ${aLabel}` : aLabel).toLowerCase();
            const bCommand = (b.category ? `${LocalizationInfo.localize(b.category)}: ${bLabel}` : bLabel).toLowerCase();
            return (aCommand).localeCompare(bCommand);
        } else {
            return 0;
        }
    }
```
The commands module, in order to implement a simple utility, needs to know that it may have to translate the input, and the same will need to be done everywhere we touch a string that might ever want to be localized - and at the limit, that's basically anywhere a string is handled.

By contrast, VSCode's solution looks like it happens at load time, and thereafter the application can behave as though it's dealing with fixed strings, but you have to reload the application to apply a new localization. In general, we can't make very many assumptions about what we know at load time, so we may not be able to achieve the same level of parsimony that VSCode has, but I wonder if there's an architectural alternative that reduces the impact relative to the current approach.

msujew commented:
Hi @colin-grant-work, thanks for the input.

I also wasn't too happy about the `label: string | LocalizationInfo` solution. I present another solution in a [fork of my `i18n` branch](https://github.com/msujew/theia/tree/i18n-express) which uses a more [vscode like API](https://github.com/msujew/theia/blob/4ca1f1a79105eae26d93c1bd79a5306eae89ec67/packages/workspace/src/browser/workspace-commands.ts#L54-L59) by loading everything on startup. However, this solution is quite diametrical towards the usual architecture of Theia, since it uses [an express endpoint](https://github.com/msujew/theia/blob/4ca1f1a79105eae26d93c1bd79a5306eae89ec67/packages/core/src/node/backend-application.ts#L211-L219) to serve localizations instead of using the websocket connection. What do you think about that solution?

paul-marechal commented:
> However, this solution is quite diametrical towards the usual architecture of Theia, since it uses [an express endpoint](https://github.com/msujew/theia/blob/4ca1f1a79105eae26d93c1bd79a5306eae89ec67/packages/core/src/node/backend-application.ts#L211-L219) to serve localizations instead of using the websocket connection.

Serving things over regular HTTP is not completely crazy if we are talking about fetching data. Remote logic with side effects are usually better encapsulated with the RPC system. So this is fine.

On the other hand, the way you defined the express route is pretty weird:
- Why not use a proper `BackendApplicationContribution` to initialize the localization registry?
- Then replace the `this.use(...)` by something like `configure(app)` and do `app.get('/i18/:locale', ...)`?

msujew commented:
@paul-marechal True, if we look at it from a pure fetching perspective, that makes more sense. I always had the additional overhead of storing the locale in the backend (for the purpose of localizing vscode extensions) in the back of my mind. However, just fetching Theias own localization is completely independent of that, so doing this through an http call would be appropriate. For everything with else we could still use the RPC based API.

And yes, I just wanted to have a quick and dirty solution for presenting the express based approach. Implementing all of that as a proper `BackendApplicationContribution` looks way better.

I'll refactor my original code and come back to you.

msujew commented:
@paul-marechal I implemented everything with a `BackendApplicationContribution` and removed the `LocalizationService` completely. Do you have an idea where to put the `loadTranslations` call (currently made [here](https://github.com/msujew/theia/blob/3543dc6e09aeacd2bad72c6deabe751463b29e39/packages/core/src/browser/frontend-application.ts#L30) to ensure that the translations are loaded before `localize` is called?

msujew commented:
@paul-marechal You said in the dev meeting you had issues getting my changes to actually work. I just pushed some fixes for another issue. However, everything seems to be working fine in the Gitpod environment I was working with. Can you give me some pointers towards your startup issue?

paul-marechal commented on packages/core/src/browser/frontend-application.ts:
```diff
@@ -25,6 +25,9 @@ import { FrontendApplicationStateService } from './frontend-application-state';
 import { WindowService } from './window/window-service';
+import { loadTranslations } from './nls';
+
+loadTranslations();
```
I don't think we should be triggering an async operation at this point and then cross fingers that the rest of the modules will properly pick up what resolved here...

Here's a fix: https://github.com/msujew/theia/compare/i18n...eclipse-theia:mp/i18n

Note that I had to modify the monaco loader as it didn't like not running after the page was loaded (due to waiting for the translation request to go through).

msujew replied:
Thanks for the fix, I applied it to my branch and it works really well.

msujew commented:
I just implemented a [vscode style](https://github.com/microsoft/vscode/blob/f84cd8f4274424f1667a694da32b885f72d0e92b/src/vs/platform/actions/common/actions.ts#L21-L30) `LocalizedString` type. We need that type to store the original untranslated name of commands to display them under the translated name in the quick input.

However, it increases maintenance efforts and is a breaking change. WDYT, is that a feature we want to support in Theia? Or do we rather use something like `originalLabel?: string, originalCategory?: string` on commands?

paul-marechal commented:
> We need that type to store the original untranslated name of commands to display them under the translated name in the quick input.

I wasn't sure what you meant, so here's how it looks like in VS Code:

![image](https://user-images.githubusercontent.com/14182238/122152120-c68ed900-ce2e-11eb-9847-91eee4e364fe.png)

msujew commented:
I actually didn't like my first approach because of the breaking changes it introduces. Therefore I already implemented the `originalLabel?: string, originalCategory?: string` approach.

colin-grant-work commented on packages/navigator/src/browser/navigator-contribution.ts:
```diff
@@ -619,7 +620,7 @@ export class FileNavigatorContribution extends AbstractViewContribution<FileNavi
             this.toDisposeAddRemoveFolderActions.push(registry.registerMenuAction(NavigatorContextMenu.WORKSPACE, {
                 commandId: FileNavigatorCommands.ADD_ROOT_FOLDER.id,
-                label: WorkspaceCommands.ADD_FOLDER.label!
+                label: WorkspaceCommands.ADD_FOLDER.label
```
In several places (e.g. `packages/editor/src/browser/editor-menu.ts`), you've added the non-null `!` to labels for menu actions. Here you're removing it. Is there any particular reason to do either?

msujew replied:
Actually no. These are just left-overs of my previous changes, which I didn't clean up.

colin-grant-work commented on packages/core/src/common/command.ts:
```diff
@@ -33,6 +33,7 @@ export interface Command {
      * A label of this command.
      */
     label?: string;
+    originalLabel?: string;
```
It might be useful to assist the transition to this system to create a utility that will generate this structure from commands that adhere to the older interface? Something like 
```typescript
export function toLocalizedCommand(command: Command, nlsLabelKey: string = command.id, nlsCategoryKey: string | undefined = command.category): Command {
    return {
        ...command,
        label: command.label && nls.localize(nlsLabelKey, command.label),
        originalLabel: command.label,
        category: nlsCategoryKey && command.category && nls.localize(nlsCategoryKey, command.category),
        originalCategory: command.category,
    };
}
```

msujew replied:
I like this, I added it. You can see its usage in the `workspace-commands.ts`.

colin-grant-work commented on packages/core/src/browser/common-frontend-contribution.ts:
```diff
@@ -987,6 +999,33 @@ export class CommonFrontendContribution implements FrontendApplicationContributi
         }
     }
 
+    protected async configureDisplayLanguage(): Promise<void> {
```
Running this in Electron and choosing the German option gave me a white window that never loaded. No untoward logs on the backend, though, so I'm not exactly sure what happened.

msujew replied:
I figured this one out. It seems like `String.prototype.replaceAll` wasn't available in the electron frontend. I used it in the translation process which led to an error on startup. Replacing it with `String.prototype.replace` and a regex fixed the issue.

colin-grant-work replied:
That seems to have fixed it. I'm now able to Configure Display Lanuguage / Sprache Wählen freely in Electron. 👍

iamtangram commented:
Hello, how about this PR progress? We are waiting for new release. haha~

msujew commented:
@iamtangram I'm currently waiting for further reviews. If you're so eager for this feature to release, you can join the discussion as well, discussing open questions that I posted in the initial PR comment or doing a review of my code ;)

Any input - even from non-commiters of Theia - is greatly appreciated.

colin-grant-work commented:
It looks like some of the manipulation either of command labels or menu labels is causing colons to show up in menus where they shouldn't:
![image](https://user-images.githubusercontent.com/62660806/123432827-69aec380-d590-11eb-8849-afc31d214b60.png)

colin-grant-work commented:
> Monaco can't be translated via vscode language packs, as all translations are done during runtime, while monaco is translated during load-time. This isn't an issue for the default languages (like de, it, fr...), but makes it impossible for vscode language packs to actually contribute to monaco's translation process.

Is this still the case, in particular, aren't all of the translations now performed at load time?

msujew commented:
> Is this still the case? In particular, aren't all of the translations now performed at load time?

@colin-grant-work I actually just finished trying to override the `loadBundle` method that monaco uses for custom translations. While it works theoretically, I don't recommend to actually support this usecase. The issue is that we have to transform the key value pairs into an array of translated messages. This is the way vscode extensions usually work: A gulp-post-processing step transforms every `localize(<key>, <default>)` into `localize(<keyIndex>, null)`. This is also the case for monaco. While transforming the key value pairs into an array is possible - even keeping the correct order - any slight mismatch within that process, due to a wrongly calculated index, leads to an index-out-of-range error and the frontend application crashes.

This becomes quite problematic due to two further issues: 
* We don't know how many values monaco expects in the array
* We need to use the exact version of the localization pack that the monaco version was released with. In other case, the keys no longer match up (due to the index calculation thing) and the application no longer starts. This basically is a side-effect from another open issue, namely how we deal with incompatible versions of language packs.

I think the language pack version question has to be solved first.

msujew commented:
> It looks like some of the manipulation either of command labels or menu labels is causing colons to show up in menus where they shouldn't

That was an oversight. I just fixed that.

misakajimmy commented:
this pr looks interesting，i want to keep on use this pr in my project until it become official.

misakajimmy commented:
wow,it looks interesting,i will keep using this until it become offical.

msujew commented:
@misakajimmy I'm glad my work is already useful to you. Consider adding any bugs/quirks/weird behavior that you notice as a review to this PR.

jbicker commented:
@msujew just rebased on top of `master` and integrated the monaco update changes.
As the review process for this PR is a bit slow at the moment (understandably), we would kindly ask the @eclipse-theia/core  team if anyone is up for a review. Thanks!

paul-marechal reviewed (commented):
For some reason, the `Configure Display Language` command does not show the original label:

![image](https://user-images.githubusercontent.com/14182238/124521956-6d77ec80-ddbf-11eb-9807-2e54fea21f2c.png)

tsmaeder commented:
@msujew I think we should have some guidance on how to use I18N in the [Coding Guidelines](https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines). How and where are contributors supposed to internationalize their code? How do we add new languages, etc.?

tsmaeder commented:
Additional documentation is needed for the message format used: how do we write message strings with parameters and what types do we support for the arguments (you mentioned dates and number and yes, they should be internationalized. If we can point to some VS Code doc for that, that's fine, but we still need to point.

msujew commented on packages/core/src/browser/nls.ts:
```diff
@@ -0,0 +1,78 @@
+
+    localization?: Localization;
+
+    addLocalization(localizations: Localization): void {
```
I actually removed the `addLocalization` method. It used to do more than just setting the publicly visible property, but now it's kind of useless.

msujew replied:
Also fixed the issue with `Configure Display Language` not showing the original translation.

msujew commented on packages/core/src/common/command.ts:
```diff
@@ -50,6 +53,17 @@ export namespace Command {
     }
 
+    /** Utility function to easily translate commands */
+    export function toLocalizedCommand(command: Command, nlsLabelKey: string = command.id, nlsCategoryKey: string | undefined = command.category): Command {
```
Didn't think about that, right. `nlsCategory?: string` is the new signature.

msujew commented:
@tsmaeder Right, makes sense. I will write something up within the next days.

Also, regarding date and number internationalization: I'm specifically asking, since vscode i18n library **only** accepts strings. Nothing else, meaning no numbers nor dates or anything else. Do we want support more than vscode does? I'd argue that's out of scope, at least for this PR.

misakajimmy commented:
I desperately want theia to support i18n, so I want to contribute some code to promote the i18n of theia. Can I contribute some PR to the i18n branch now?

misakajimmy commented:
I found that after using this i18n solution, the order of the drop-down menu may be disordered in some languages. Is there a way to not change the order or sort it?

msujew commented:
@misakajimmy You mean the command quick pick? Theoretically, we could change the `Command.compare` method to always compare via the original label. But that's not the way it's done in vscode and we usually strive to align to vscode behavior. vscode will always sort by the label of the current locale and not the original english version.

What kind of contributions do you mean? Improvements to the code should be done as a review (but you can send me a PR yourself by using a link in the following fashion: `https://github.com/msujew/theia/compare/i18n...misakajimmy:<branchname>`). Additional changes like translations for other commands, menus, etc. should probably be done in a separate PR to the Theia repo. This PR just adds the basic functionality, actual content is out of scope, except for the translations I did for testing purposes.

tsmaeder commented:
> Do we want support more than vscode does? I'd argue that's out of scope, at least for this PR.

If we can do better without breaking plugin API compatibility, I don't see why not. Or do you want to see "Ihr Computer startet um 12pm neu".  But yeah, doesn't have to be in this PR.

tsmaeder commented:
@misakajimmy commands being ordered according to the localized label seems right to me. 

You are very welcome to contribute, but it doesn't necessarily have to be in this PR.

msujew commented:
> How and where are contributors supposed to internationalize their code? How do we add new languages, etc.?

@tsmaeder I just spend a bit of time thinking about this. In my initial naivety I thought we could leverage the community to contribute translations that are not part of the vscode language packs, but I don't think that's feasible for 10+ supported languages. What do you think about using a service like [transifex](https://www.transifex.com/) to automatically translate any "original theia" content? We would probably have to add another dev-package to somehow extract the keys and default values of translations like [vscode does](https://github.com/Microsoft/vscode-nls-dev).

tsmaeder commented:
> What do you think about using a service like transifex to automatically translate any "original theia" content?

As I don't know the service, it's hard to tell. Do you want to open a feature request detailing what you want to do?

msujew commented:
> As I don't know the service, it's hard to tell. Do you want to open a feature request detailing what you want to do?

Sure, I just created #9708 where we can discuss anything related to that.

marcdumais-work commented:
At a quick glance, I can't easily tell what's pending before this PR can be approved. @tsmaeder do we need #9708 to be resolved before we can proceed? What else needs addressing?

marcdumais-work commented:
@msujew I see there are currently merge conflicts - it could be a good idea to rebase the PR on latest master?

marcdumais-work commented:
> (but you can send me a PR yourself by using a link in the following fashion: https://github.com/msujew/theia/compare/i18n...misakajimmy:<branchname>)

We need be careful about integrating an other person's changes within a pull-request (*) - any contribution should be done "above board", without obfuscating the provenance/authorship, and are subject to the usual Eclipse Foundation rules. For example:

_When the contributor has signed the Eclipse Contributor Agreement (ECA) and the following conditions are met, a CQ is not required:_
 - _Was developed from scratch; written 100% by submitting contributor;_

See [here](https://www.eclipse.org/projects/handbook/#ip-project-content) for more details.

msujew commented on packages/core/src/node/i18n/localization-contribution.ts:
```diff
@@ -0,0 +1,103 @@
+    protected readonly contributions: ContributionProvider<LocalizationContribution>;
+
+    async initialize(): Promise<void> {
+        for (const contrib of this.contributions.getContributions()) {
+            await contrib.registerLocalizations(this);
+        }
```
Yes, concurrency issues shouldn't come up here. 👍

msujew commented on packages/plugin-ext/src/hosted/node/plugin-service.ts:
```diff
@@ -113,4 +119,35 @@ export class HostedPluginServerImpl implements HostedPluginServer {
     }
+
+    protected async localizePlugin(plugin: DeployedPlugin, locale: string): Promise<DeployedPlugin> {
+        plugin = { ...plugin };
```
No it doesn't. It also not necessary to copy the element at all, since even if we modify the original object we will re-modify it, as soon as we reload it. Though it may be a code smell to modify it at all.

msujew commented on packages/core/src/node/i18n/localization-contribution.ts:
```diff
@@ -0,0 +1,103 @@
+    protected calculateFilePath(name: string): string {
+        return path.join(__dirname, '..', '..', '..', 'i18n', name);
+    }
+}
```
Yes, requiring the JSON file sounds good. I implemented it like that. The solution becomes a lot slimmer.

msujew commented on packages/plugin-ext/src/hosted/node/plugin-manifest-loader.ts:
```diff
@@ -36,14 +36,18 @@ export async function loadManifest(pluginPath: string): Promise<any> {
+            if (e.code !== 'ENOENT') {
+                throw e;
+            }
+            return {};
```
I changed it to look for the localized file first and only load it, if it actually exists. This removes the need for nested try catch blocks.

msujew commented:
@tsmaeder I wrote a small proposal for the coding guidelines:

## I18n

1. Always translate any user facing text with the `nls.localize(key, defaultValue, ...args)` function.
> What is user facing text? Any strings that are hardcoded (not calculated) that could be in any way visible to the user, be it labels for commands and menus or messages. 

1.1. Parameters for messages should be passed as the `args` of the `localize` function. They are inserted at the location of the placeholders - in the form of `{\d+}` - in the localized text. E.g. `{0}` will be replaced with the first `arg`, `{1}` with the second, etc.

2. Use utility functions where possible:
```ts
// bad
command: Command = { label: nls.localize(key, defaultValue), originalLabel: defaultValue };
// good
command = Command.toLocalizedCommand({ id: key, label: defaultValue });
```

jbicker reviewed (approved):
LGTM :-)
Since @msujew's last requests have been resolved and no more have been received for about two weeks, I think it's okay to approve it. If there are no further objections, we will merge it next week.

msujew commented:
FYI, I will drop the german custom translation before merging this PR and follow up with another PR that includes all translations contributed by the vscode language packs. Or would you rather see this as a part of this PR? We would like to include everything in the August release if possible.

fstasi commented:
> FYI, I will drop the german custom translation before merging this PR and follow up with another PR that includes all translations contributed by the vscode language packs. Or would you rather see this as a part of this PR? We would like to include everything in the August release if possible.

@msujew I agree on having two separate PRs, one for the actual code implementation and another for the translations. This will reduce the noise in this one.

paul-marechal commented:
> - Language Packs can't be uninstalled without restarting the server (all localizations are held in memory after startup), as we don't know which translations are contributed by vscode language packs and which by Theia extensions **yet**.

Can you point to the code that is the cause of loading everything in memory?

> - Currently the `LocalizationProvider` holds a single active language for the whole application. Is there a way to hold a value for every user?

Sounds like state that should live in the frontend only?

> - Theia's quick-input-bar doesn't filter by the details of the item, unlike vscode. This makes finding commands quite difficult after switching the locale.

I was about to comment on this. If the plan is to not support this yet then OK to merge.

paul-marechal reviewed (approved):
I follow the "How to test" steps and everything worked.

Code LGTM, although regarding the various pending issues I am fine with merging this PR before the upcoming release if we are ready to potentially make breaking changes, as required to add/fix the rest later.

msujew commented:
> Can you point to the code that is the cause of loading everything in memory?

We load everthing here:

https://github.com/eclipse-theia/theia/blob/44fec6ff74588bd8b7a645d145cc2a693a5ac8ca/packages/plugin-ext/src/hosted/node/hosted-plugin-deployer-handler.ts#L137-L139

> Sounds like state that should live in the frontend only?

I agree. There's two issues with that though:
1. The backend needs to know the selected language by the user to start the plugin host with the correct language (for vscode extensions to localize themselves correctly)
2. The backend will regularly send stuff like messages and so on to the frontend. This will have to be translated as well.

We can work around both points I guess. For 1. we only need to know the selected language on startup, and forgot about it afterwards. This is basically how it works already, so we don't have to change anything. For 2. we would need to only send the translation keys to the frontend, which will then translate the message on its own. This is quite unintuitive though.

> I was about to comment on this. If the plan is to not support this yet then OK to merge.

This was fixed by #9928. I'll rebase everything, and then that'll work out of the box.

msujew commented:
@paul-marechal After rebasing filtering by the original command name is now possible. The only command still getting translated is the `Configure Display Language` command. You can test it that way.
